File: problem_set_2/valid_parentheses.py


def isValid(s: str) -> bool:
    """
    This functions checks whether the parentheses are valid or not
    """
    database = {
        '(': ')',
        '{': '}',
        '[': ']',
    }

    stack = []

    if len(s) <= 1:
        return False

    for character in s:
        if character in database.keys():
            stack.append(character)
        elif stack and database[stack[-1]] == character:
            stack.pop()
        else:
            return False

    if len(stack) > 0:
        return False

    return True

if __name__ == '__main__':

    print(isValid('()[]{}'))
    print(isValid('([)]'))
    print(isValid('()'))

File: problem_set_2/valid_parentheses.go
package main

import (
	"fmt"
)

func isValid(s string) bool {
	database := map[rune]rune{
		'(': ')',
		'{': '}',
		'[': ']',
	}

	stack := []rune{}

	if len(s) <= 1 {
		return false
	}

	for _, char := range s {
		if _, exists := database[char]; exists {
			stack = append(stack, char)
		} else if len(stack) > 0 && database[stack[len(stack)-1]] == char {
			stack = stack[:len(stack)-1]
		} else {
			return false
		}
	}

	return len(stack) == 0
}

func main() {
	result := isValid("()[]{}")
	fmt.Println(result) // Output: true

	result = isValid("(]")
	fmt.Println(result) // Output: false

	result = isValid("([)]")
	fmt.Println(result) // Output: false

	result = isValid("{[]}")
	fmt.Println(result) // Output: true
}
